feat: add Tagged for composable tagged types

ethanresnick opened the pull request:
This adds two new types, `Tagged` and `UnwrapTagged`, per https://github.com/sindresorhus/type-fest/issues/665.

It also tweaks `UnwrapOpaque` so that it works on types produced by `Tagged` (as well as still working on types produced by `Opaque`, of course). This is a small gesture at interoperability between `Opaque` and `Tagged`, to make it easier for codebases to migrate from `Opaque` to `Tagged`.

Fixes #665

sindresorhus commented on source/opaque.d.ts:
```diff
@@ -1,17 +1,25 @@
 	readonly [tag]: Token;
 };
 
+type MultiTagContainer<Token extends string | number | symbol> = {
```
```suggestion
type MultiTagContainer<Token extends PropertyKey> = {
```

sindresorhus commented:
This looks very good 👍

sindresorhus commented:
You need to also add them to the readme.

sindresorhus commented:
Some types like https://github.com/sindresorhus/type-fest/blob/main/source/exact.d.ts, https://github.com/sindresorhus/type-fest/blob/main/source/invariant-of.d.ts, and https://github.com/sindresorhus/type-fest/blob/main/source/conditional-pick-deep.d.ts have special handling of `Opaque`. That should probably extend to `Tagged` too.

ethanresnick commented:
Ok, I think I addressed all the comments. For the three types you mentioned that use `Opaque` (`Exact`, `InvariantOf`, and `ConditionalPickDeep`), I did something different in each case:

- `Exact`: I extended the current approach to work with `Tagged` types too. As I mentioned in the inline code comment, I'm not sure the current approach is exactly correct/ideal, but that can be a topic for a separate PR.

- `InvariantOf`: I kept the same basic logic, but inlined it to remove the dependency on `Opaque`, in case we want to deprecate `Opaque` at some point. I also put the tag that's enforcing the invariance in a differently-named symbol, so that it won't conflict with tags added by `Tagged`/`Opaque`. (Before, it was impossible to do `Opaque<InvariantOf<X>>` or `InvariantOf<Opaque<X>>`, as both `InvariantOf` and `Opaque` would've tried to put conflicting types at the same phantom `[tag]` key.)

> `ConditionalPickDeep` didn't seem like it should be using opaque/tagged types at all; from my reading, it looks like it really just wanted a `unique symbol` type. So, I updated that accordingly, and all the tests still pass.

sindresorhus commented:
Thanks ;)
